### PR TITLE
Replace C-extension with pure Python code for QCP superimposition.

### PR DESCRIPTION
**Pull request by JoaoRodrigues**
- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

This PR replaces the QCPSuperimposer C module with a pure Python implementation, with minimal loss of performance. In fact, it improves performance in my tests by ~30%. The API is largely the same, with the addition of `set_atoms` and `apply`, to mirror the behavior of `Superimposer` and simplify access to the algorithm. Also refactored the tests to use `assertEqual` instead of multiple calls to `numpy.array_equal`, which were making debugging failing tests quite difficult.

### CODE REVIEW

File: Bio/PDB/qcprot.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                q2 = -a31 * a1324_1423 + a33 * a1124_1421 - a34 * a1123_1321
+                q3 = a31 * a1224_1422 - a32 * a1124_1421 + a34 * a1122_1221
+                q4 = -a31 * a1223_1322 + a32 * a1123_1321 - a33 * a1122_1221
+                qsqr = q1 * q1 + q2 * q2 + q3 * q3 + q4 * q4
+
+                if qsqr < evecprec:
+                    rot = np.eye(3)
+                    return rmsd, rot, [q1, q2, q3, q4]
+
+    normq = math.sqrt(qsqr)
```
Comment (joaomcteixeira) [inline Bio/PDB/qcprot.py:185]: `normq = qsqr ** 0.5`

File: Bio/PDB/qcprot.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    z2 = q4 * q4
+
+    xy = q2 * q3
+    az = q1 * q4
+    zx = q4 * q2
+    ay = q1 * q3
+    yz = q3 * q4
+    ax = q1 * q2
+
+    rot = np.zeros((3, 3))
```
Comment (joaomcteixeira) [inline Bio/PDB/qcprot.py:203]: `rot` is only created to be transposed, why not creating it already transposed?

```python
rot = np.zeros((3, 3))
rot[0][0] = a2 + x2 - y2 - z2
rot[1][0] = 2 * (xy + az)
rot[2][0] = 2 * (zx - ay)
rot[0][1] = 2 * (xy - az)
rot[1][1] = a2 - x2 + y2 - z2
rot[2][1] = 2 * (yz + ax)
rot[0][2] = 2 * (zx + ay)
rot[1][2] = 2 * (yz - ax)
rot[2][2] = a2 - x2 - y2 + z2
```

File: Bio/PDB/qcprot.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    rot[0][1] = 2 * (xy + az)
+    rot[0][2] = 2 * (zx - ay)
+    rot[1][0] = 2 * (xy - az)
+    rot[1][1] = a2 - x2 + y2 - z2
+    rot[1][2] = 2 * (yz + ax)
+    rot[2][0] = 2 * (zx + ay)
+    rot[2][1] = 2 * (yz - ax)
+    rot[2][2] = a2 - x2 - y2 + z2
+
+    return rmsd, rot.T, [q1, q2, q3, q4]
```
Comment (joaomcteixeira) [inline Bio/PDB/qcprot.py:214]: would return a tuple instead `(q1, q2, q3, q4)`.

File: Bio/PDB/qcprot.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        delta = (a * mxEigenV + C0) / (2.0 * x2 * mxEigenV + b + a)
+        mxEigenV -= delta
+        if abs(mxEigenV - oldg) < abs(evalprec * mxEigenV):
+            break
+    else:
+        print(f"Newton-Rhapson did not converge after {nr_it} iterations")
+
+    # The original code has a guard if minScore > 0 and rmsd < minScore, although
+    # the default value of minScore is -1. For simplicity, we ignore that check.
+    rmsd = math.sqrt(2.0 * abs(E0 - mxEigenV) / natoms)
```
Comment (joaomcteixeira) [inline Bio/PDB/qcprot.py:121]:
```python
rmsd = (2.0 * abs(E0 - mxEigenV) / natoms) ** 0.5
```

**Comment by joaomcteixeira** [comment]
Crazy implementation :wink:
Gave some very minor comments :+1:

**Comment by JoaoRodrigues** [comment]
@peterjc any idea on why the tests would be failing with the import errors? Locally they all pass (on MacOS and Windows).

**Comment by peterjc** [comment]
e.g.

``ERROR while importing Bio.pairwise2: cannot import name '_aligners' from partially initialized module 'Bio.Align' (most likely due to a circular import) (D:\a\biopython\biopython\Bio\Align\__init__.py)``

I have no idea, @mdehoon may be able to tell? There is nothing obviously changed on the C side in this PR that I can see.

**Comment by JoaoRodrigues** [comment]
Thanks. This is also happening on master, the latest commit is also failing with this message. The weird thing is that it seems to run fine on Linux, so maybe it's a CI configuration thing with MacOS and Windows?

**Comment by JoaoRodrigues** [comment]
I can reproduce this locally with `python=3.7` but none of the others. Working on a fix now.

**Pull request merged**